File: android_app_sample/android_sample/src/main/java/com/gaa/iap/sample/widget/ResultNumberView.java
package com.gaa.iap.sample.widget;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.gaa.iap.sample.R;

import java.util.ArrayList;
import java.util.List;

public class ResultNumberView extends ConstraintLayout {

    private RecyclerView mRecyclerView;

    public ResultNumberView(Context context) {
        super(context);
        init();
    }

    public ResultNumberView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        final Context context = getContext();
        View view = LayoutInflater.from(context).inflate(R.layout.view_result, this, true);
        mRecyclerView = view.findViewById(R.id.recycler);
        mRecyclerView.addItemDecoration(new DividerItemDecoration(context));
        mRecyclerView.setLayoutManager(new LinearLayoutManager(context));
//        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setOverScrollMode(View.OVER_SCROLL_NEVER);
    }

    public void setData(List<Integer> luckyNumbers, List<List<Integer>> myNumbers) {
        mRecyclerView.setAdapter(new BallAdapter(luckyNumbers, myNumbers));
    }

    public void clear() {
        if (mRecyclerView.getAdapter() != null) {
            int count = mRecyclerView.getAdapter().getItemCount();
            mRecyclerView.getAdapter().notifyItemRangeRemoved(0, count);
        }
    }

    private static class BallAdapter extends RecyclerView.Adapter<BallViewHolder> {

        private List<Integer> mLuckyNumbers;
        private List<List<Integer>> mMyNumbers;

        private BallAdapter(List<Integer> luckyNumbers, List<List<Integer>> myNumbers) {
            mLuckyNumbers = luckyNumbers;
            mMyNumbers = myNumbers;
        }

        @NonNull
        @Override
        public BallViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(parent.getContext());
            return new BallViewHolder(inflater.inflate(R.layout.view_result_item, parent, false));
        }

        @Override
        public void onBindViewHolder(@NonNull BallViewHolder holder, int position) {
            holder.setNumber(mLuckyNumbers, mMyNumbers.get(position), position);
        }

        @Override
        public int getItemCount() {
            return mMyNumbers.size();
        }
    }

    private static final class BallViewHolder extends RecyclerView.ViewHolder {
        private TextView mRankView;
        private ImageView mBonusView;
        private ArrayList<BallView> mBallViews = new ArrayList<>();

        private BallViewHolder(@NonNull View itemView) {
            super(itemView);
            mRankView = itemView.findViewById(R.id.rank);
            mBonusView = itemView.findViewById(R.id.bonus);
            ViewGroup group = (ViewGroup) itemView;
            for (int i = 1; i < group.getChildCount(); i++) {
                View child = group.getChildAt(i);
                if (child instanceof BallView) {
                    mBallViews.add((BallView) child);
                }
            }
        }

        void setNumber(List<Integer> luckyNums, List<Integer> myNums, int position) {
            int luckyCount = 0;
            for (int i = 0; i < myNums.size(); i++) {
                boolean isLucy = false;
                int myNum = myNums.get(i);
                for (int j = 0; j < luckyNums.size(); j++) {
                    if (myNum == luckyNums.get(j)) {
                        luckyCount++;
                        isLucy = true;
                        break;
                    }
                }
                mBallViews.get(i).setNumber(myNum, isLucy);
            }

            mRankView.setText(getRankString(luckyCount));
            mBonusView.setVisibility(position >= 5 ? VISIBLE : GONE);
        }

        private String getRankString(int number) {
            switch (number) {
                case 3: return "4th";
                case 4: return "3rd";
                case 5: return "2nd";
                case 6: return "1st";
                default: return " - ";
            }
        }
    }


    private static class DividerItemDecoration extends RecyclerView.ItemDecoration {
        private Drawable divider;

        private DividerItemDecoration(Context context) {
            final TypedArray styledAttributes = context.obtainStyledAttributes(new int[]{android.R.attr.listDivider});
            divider = styledAttributes.getDrawable(0);
            styledAttributes.recycle();
        }

        @Override
        public void onDraw(@NonNull Canvas canvas, RecyclerView parent, @NonNull RecyclerView.State state) {
            int left = parent.getPaddingLeft();
            int right = parent.getWidth() - parent.getPaddingRight();
            int childCount = parent.getChildCount();
            for (int i = 0; i < childCount; i++) {
                View child = parent.getChildAt(i);
                RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
                int top = child.getBottom() + params.bottomMargin;
                int bottom = top + divider.getIntrinsicHeight();
                divider.setBounds(left, top, right, bottom);
                divider.draw(canvas);
            }
        }
    }
}

File: onestore_iap_sample/sample_luckyone/src/main/java/com/onestore/sample/inapp/widget/ResultNumberView.kt
package com.onestore.sample.inapp.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.onestore.sample.inapp.R

class ResultNumberView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int = 0)
    : ConstraintLayout(context, attrs, defStyle) {

    private var mRecyclerView: RecyclerView

    init {
        val view: View = LayoutInflater.from(context).inflate(R.layout.view_result, this, true)
        mRecyclerView = view.findViewById(R.id.recycler)

        with(mRecyclerView) {
            addItemDecoration(DividerItemDecoration(context))
            layoutManager = LinearLayoutManager(context)
            overScrollMode = OVER_SCROLL_NEVER
        }
    }

    fun setData(luckyNumbers: List<Int>, myNumbers: List<List<Int>>) {
        mRecyclerView.adapter = BallAdapter(luckyNumbers, myNumbers)
    }

    fun clear() {
        val adapter = mRecyclerView.adapter
        if (adapter != null) {
            val count = adapter.itemCount
            adapter.notifyItemRangeRemoved(0, count)
        }
    }

    class BallAdapter constructor(private val mLuckyNumbers: List<Int>, private val mMyNumbers: List<List<Int>>)
        : RecyclerView.Adapter<BallViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BallViewHolder {
            val inflater = LayoutInflater.from(parent.context)
            return BallViewHolder(inflater.inflate(R.layout.view_result_item, parent, false))
        }

        override fun onBindViewHolder(holder: BallViewHolder, position: Int) {
            holder.setNumber(mLuckyNumbers, mMyNumbers[position])
        }

        override fun getItemCount(): Int {
            return mMyNumbers.size
        }
    }

    class BallViewHolder constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val mRankView: TextView = itemView.findViewById(R.id.rank)
        private val mBallViews = ArrayList<BallView>()

        fun setNumber(luckyNums: List<Int>, myNums: List<Int>) {
            var luckyCount = 0
            for ((index, myNum) in myNums.withIndex()) {
                var isLucy = false
                for (lucky in luckyNums) {
                    if (myNum == lucky) {
                        luckyCount++
                        isLucy = true
                        break
                    }
                }
                mBallViews[index].setNumber(myNum, isLucy)
            }
            mRankView.text = getRankString(luckyCount)
        }

        private fun getRankString(number: Int): String {
            return when (number) {
                3 -> "4th"
                4 -> "3rd"
                5 -> "2nd"
                6 -> "1st"
                else -> " - "
            }
        }

        init {
            val group = itemView as ViewGroup
            for (i in 1 until group.childCount) {
                val child = group.getChildAt(i)
                if (child is BallView) {
                    mBallViews.add(child)
                }
            }
        }
    }

    class DividerItemDecoration constructor(context: Context)
        : RecyclerView.ItemDecoration() {

        private val divider: Drawable?

        init {
            val styledAttributes = context.obtainStyledAttributes(intArrayOf(android.R.attr.listDivider))
            divider = styledAttributes.getDrawable(0)
            styledAttributes.recycle()
        }

        override fun onDraw(canvas: Canvas, parent: RecyclerView, state: RecyclerView.State) {
            if (divider != null) {
                val left = parent.paddingLeft
                val right = parent.width - parent.paddingRight
                val childCount = parent.childCount

                for (i in 0 until childCount) {
                    val child = parent.getChildAt(i)
                    val params = child.layoutParams as RecyclerView.LayoutParams
                    val top = child.bottom + params.bottomMargin
                    val bottom = top + divider.intrinsicHeight

                    divider.setBounds(left, top, right, bottom)
                    divider.draw(canvas)
                }
            }
        }
    }
}